### ISSUE TITLE
'Test Failure' reported as 'Success' in TeamCity when message contains Unicode ’ char (RIGHT SINGLE QUOTATION MARK)

### ISSUE THREAD

**Issue by slawomir-brzezinski-at-interxion** [xunit/xunit#1185]
Using xUnit 2.2 and TeamCity 10.0.4 on Windows.

## Reproduction
Please see [the reproduction repo](https://github.com/slawomir-brzezinski-at-interxion/xUnitInTeamCity.ReproForUnicodeEscapingIssue). It contains [one test](https://github.com/slawomir-brzezinski-at-interxion/xUnitInTeamCity.ReproForUnicodeEscapingIssue/blob/master/xUnitInTeamCity.ReproForUnicodeEscapingIssue.Tests/IssueReproTests.cs#L9).

1. Point TeamCity at the repo
1. Add a 'Nuget Restore step' in the beginning
1. Add an off-the-shelf xUnit runner step with '**/*.Tests.dll'
1. Observe:
Despite the runner exits with non-zero code (thus correctly failing the build), the TeamCity report **shows the test itself as successful**.

## Analysis
Inspecting the log shows a message:
```
##teamcity[testFailed name='xUnitInTeamCity.ReproForUnicodeEscapingIssue.Tests.IssueReproTests.FailWithMessageContainingNonAsciiChars' details='Assert.Equal() Failure|r|n            (pos 0)|r|nExpected: ' +Ø%D|r|nActual:   anything different so that test fails|r|n            (pos 0)|r|n   at xUnitInTeamCity.ReproForUnicodeEscapingIssue.Tests.IssueReproTests.FailWithMessageContainingNonAsciiChars() in d:\Teamcity\BuildAgent\work\c80283b4cb57ca79\xUnitInTeamCity.ReproForUnicodeEscapingIssue.Tests\IssueReproTests.cs:line 25' flowId='4862240aeba54317941ec7648b984516']
Incorrect property name.
Valid property list format is (name( )*=( )*'escaped_value'( )*)* where escape symbol is "|"
```

As noted in the [repro source](https://github.com/slawomir-brzezinski-at-interxion/xUnitInTeamCity.ReproForUnicodeEscapingIssue/blob/master/xUnitInTeamCity.ReproForUnicodeEscapingIssue.Tests/IssueReproTests.cs#L9), we have traced this down to the problem of console encoding, which causes the special characters to be changed to their 'simplified non-Unicode' version upon output. Especially the ’ (RIGHT SINGLE QUOTATION MARK) character is of note, because it gets changed to simple ', which gets to TeamCity unescaped and prevents it from receiving the test failure command.

### Severity
As for severity, while thanks to the non-zero exit code the developers should notice the problem, some wrapping scripts may sometimes not pass that to the end, making the failure completely unnoticed (notably [xUnit.net-dotCover](https://github.com/JetBrains/meta-runner-power-pack/tree/master/xUnit.net-dotCover) from meta-runner-power-pack [has this issue](https://github.com/JetBrains/meta-runner-power-pack/issues/80) as of present).

## Fix suggestion
Even though [some solutions may exist](http://stackoverflow.com/questions/5750203/how-to-write-unicode-characters-to-the-console) which make the pipeline not change the characters, I think that it might be good to implement a fix that works out-of-the-box in any setup. That is, just use the [TeamCity's \uNNNN escaping](https://confluence.jetbrains.com/display/TCD10/Build+Script+Interaction+with+TeamCity#BuildScriptInteractionwithTeamCity-Escapedvalues) for any special characters beyond ASCII. After all, these commands are not seen by human in a normal run. The extra benefit will be that all other special characters will start showing in TeamCity without getting changed.

Take note there's even a similar [Pull Request to in JetBrains' utility library for outputting the commands](https://github.com/JetBrains/TeamCity.ServiceMessages/pull/7/files)  that also suggest always doing it under the hood.

**Comment by hboisselle** [comment]
@slawomir-brzezinski-at-interxion Have you found any workaround? I've got the same issue here.

**Comment by slawomir-brzezinski-at-interxion** [comment]
@hboisselle well we just replaced these special quotation marks with ' for the time being (they get properly escaped), but still await some acknowledgement/answer here, as to how this is planned to be fixed, which would give us (or the community) a green light to start investing time into a PR.

**Comment by hboisselle** [comment]
@slawomir-brzezinski-at-interxion I'm working onto a fix since it is not an option in my case to replace all occurences of `'`. I would propose this quick fix in TeamCityReporterMessageHandler:
` 
static string Escape(string value)
        {
            if (value == null)
                return string.Empty;

            return value.Replace("|", "||")
                        .Replace("'", "|'")
                        .Replace("\u2019", "|'")
                        .Replace("\u2018", "|'")
                        .Replace("\r", "|r")
                        .Replace("\n", "|n")
                        .Replace("]", "|]")
                        .Replace("[", "|[")
                        .Replace("\u0085", "|x")
                        .Replace("\u2028", "|l")
                        .Replace("\u2029", "|p");
        }
 `

I know this is a very specific fix, I'd like to know what contributors would propose instead.

**Comment by slawomir-brzezinski-at-interxion** [comment]
If I were to add that encoding I'd do it something like:

```
        public static string EscapeForTcServiceMessage(this string stringValue)
        {
            stringValue = stringValue.Replace("|", "||").Replace("\'", "|\'").Replace("\n", "|n").Replace("\r", "|r").Replace("[", "|[").Replace("]", "|]");
            return GetStringWithDiacriticsEscapedAsUnicode(stringValue);
        }

        private static bool IsAscii(char ch) => ch <= '\x007f';

        private static string GetStringWithDiacriticsEscapedAsUnicode(string stringValue)
        {
            var sb = new StringBuilder();
            foreach (var charValue in stringValue)
            {
                if (BuildScriptIntegrationExtensions.IsAscii(charValue) == false)
                    sb.Append("|0x" + ((int)charValue).ToString("X4"));
                else
                    sb.Append(charValue);
            }
            return sb.ToString();
        }
```

The extra benefit is that all other international characters would also be propagated correctly to TeamCity (they currently aren't, so my Polish ł gets there as l).

**Comment by hboisselle** [comment]
@slawomir-brzezinski-at-interxion That's a nice fix. It should work all the way to TeamCity's output. I think you should make a pull request and see what others think about that.

**Comment by slawomir-brzezinski-at-interxion** [comment]
@hboisselle I updated the code slightly for performance.

That was relatively quick as I had the above code handy, but It might not be easy for us to get time to craft a proper PR too quickly, so feel free to prepare one if you are in a need for something earlier :)

**Comment by slawomir-brzezinski-at-interxion** [comment]
Fixed by #1578
